### ISSUE TITLE
Minimum supported Mono version

### ISSUE THREAD

**Issue by akoeplinger** [Pash-Project/Pash#334]
I'm wondering which Mono version you're targeting as the minimum support version. From your Travis build it looks like you're compatible with Mono 2.10 as it ships in Ubuntu 12.04. Does using this old version cause you any troubles?

Xamarin has proper [Linux packages](http://www.mono-project.com/docs/getting-started/install/linux/) for the latest Mono versions now and I guess it'd save you troubles not needing to stay compatible with 2.10 in the future as Mono matures.

**Comment by ygra** [comment]
I think the main reason we're currently on that minimum version is that Travis builds on Ubuntu 12.04 and we just use the stock Mono there (which admittedly is also easier than trying to instruct users on how to obtain another version).

There are a handful of workarounds around Mono bugs in the code, but nothing I'm aware of that would impact the experience of *using* Pash in any way.

**Comment by sburnicki** [comment]
Exactly, we just support Mono 2.10 because "why not" ;) I guess supporting old versions is a good thing if they don't make too much trouble.

When we're getting to the point when we really need new mono features we could consider to leave these old versions behind. For now Pash simply runs also on outdated systems :)

**Comment by akoeplinger** [comment]
Sure, I'm all for supporting older versions when it's not too much hassle :)

The reason I asked was that I'm one of the contributors to the [Travis CI C# support](http://blog.travis-ci.com/2014-12-10-community-driven-language-support-comes-to-travis-ci/) and looked at what projects might be interested in using it (we install the latest Mono from Xamarin's repos).

**Comment by sburnicki** [comment]
This is nice to hear, so we know we know we won't have trouble with Travis if we decide to increase the minimum target version one day :)

**Comment by akoeplinger** [comment]
Yep, just ping me if you have questions. I guess my question is answered, so I'm closing this.

**Comment by JayBazuzi** [comment]
@akoeplinger - Yes, using Mono 2.10 does cause troubles, and I would like us to upgrade as soon as responsible. There's a bug in old Mono that we work around + we are holding off on NuGet because of this.

However, we must also be mindful of the platforms that people are using w/ Pash. We don't want to force people to install a bunch of prerequisites (e.g. a newer Mono). What is a reasonable expectation these days?

**Comment by akoeplinger** [comment]
@JayBazuzi well, Xamarin offers .deb and .rpm packages, so that should cover Debian, Ubuntu, CentOS, Fedora, SUSE and their derivatives. The only thing you need to do as a user is to add the Xamarin repository as shown in the [install guide](http://www.mono-project.com/docs/getting-started/install/linux/).

Of course that's one additional step compared to just installing Mono from the distributions' repository, but honestly anyone using Mono for more than just a simple demo would want the latest version anyway IMO.

**Comment by akoeplinger** [comment]
FYI, the build with Mono 3.10 via Travis C# support seems to work fine: https://travis-ci.org/akoeplinger/Pash

**Comment by JayBazuzi** [comment]
Thanks for running that test. Good to know we are green.

When this last came up, the other Pash contributors said they like the idea of using stock Mono only. I wonder if that is still true?

I've been living in a Windows world for a while now, so I haven't looked at what recent Linux distributions are doing with Mono.

**Comment by JayBazuzi** [comment]
Related:https://github.com/Pash-Project/Pash/issues/29

**Comment by JayBazuzi** [comment]
Ubuntu 14.04 LTS comes with Mono 3.2.8, but as long as Travis is on 12.04, I think we want to keep that as our primary supported platform.

**Comment by castarco** [comment]
Hi. I think that depending on Mono > 3.2.x it's not a good idea because it's not supported in Ubuntu 14.04 (a LTS version) and because last Mono's installer for Windows it's also on the 3.2.3 version.

I'm also using Pash in Wine for some experiments... ( https://appdb.winehq.org/objectManager.php?sClass=application&iId=16572 ) ^^U, so I need support for that version of Mono.

**Comment by JayBazuzi** [comment]
@akoeplinger would it be possible to make Travis CI for C# use Mono 3.2.8, since that's what's in the last Ubuntu LTS?

**Comment by akoeplinger** [comment]
@JayBazuzi we're talking about allowing older Mono versions over at https://github.com/travis-ci/travis-ci/issues/3018. I think we'll be able to provide Mono 2.10 (as it ships on Ubuntu 12.04) and Mono 3.2.8 (via a backporting PPA). Versions starting from 3.8+ are available in the official Xamarin repos. I think that should cover most use cases.

@castarco just FYI, the upcoming Mono 3.12 release will have a reworked and updated Windows installer.

**Comment by castarco** [comment]
@akoeplinger  Yes. It happened today :+1: .

By the way, I was experimenting with Mono support in Travis CI, and I've discovered that it can be combined with Coverity Scan. I think this would be interesting to find issues in the code.

**Comment by JayBazuzi** [comment]
Neat. I'm curious what coverity woud say about pash. I like safety nets.

**Comment by ygra** [comment]
@castarco, @JayBazuzi: See #250, I did a run with Coverity a while ago. There were only a few issues, mostly either resource leaks or potential `null` dereferences. As someone who has used various static analysis tools on large C++ codebases this almost came as a disappointment how few issues C# code has in comparison ;)

**Comment by akoeplinger** [comment]
We now supports testing against multiple Mono versions including 2.10.8 and 3.2.8 (see a [test build](https://travis-ci.org/akoeplinger/Pash/builds/55689427) of Pash), I sent a PR: #359
